waiter/src/waiter/process_request.clj

```diff
@@ -180,17 +183,140 @@
     (deliver reservation-status-promise promise-value)
     (utils/exception->response (ex-info message (assoc metrics-map :status status) error) request)))
 
+(defn make-stream-reader-executor
+  "Creates and returns the ThreadPoolExecutor used for reading the input stream."
+  [concurrency-level keep-alive-mins queue-limit]
+  (let [queue (LinkedBlockingQueue. ^int queue-limit)
+        executor (ThreadPoolExecutor. concurrency-level concurrency-level keep-alive-mins TimeUnit/MINUTES queue)]
+    (metrics/waiter-gauge #(.getActiveCount executor)
+                          "core" "stream-reader" "active-thread-count")
+    (metrics/waiter-gauge #(- concurrency-level (.getActiveCount executor))
+                          "core" "stream-reader" "available-thread-count")
+    (metrics/waiter-gauge #(.getMaximumPoolSize executor)
+                          "core" "stream-reader" "max-thread-count")
+    (metrics/waiter-gauge #(-> executor .getQueue .size)
+                          "core" "stream-reader" "pending-task-count")
+    (metrics/waiter-gauge #(.getTaskCount executor)
+                          "core" "stream-reader" "scheduled-task-count")
+    executor))
+
+(defn- submit-request-streaming-task
+  "Submits a task to the executor and invokes the error handler with custom error message if the submission fails."
+  [^ThreadPoolExecutor executor runnable-task]
+  (try
+    (.execute executor runnable-task)
+    (catch Throwable throwable
+      (log/error "unable to submit task to queue")
+      (throw (ex-info "Too many concurrent requests on router"
```

Comment:
Is it possible to write an integration test for this?

Reply:
It will be hard to simulate such large numbers of concurrent requests.

Regarding integration tests as a whole, this is mainly a refactoring task and existing integration tests, including streaming tests, all pass without modification.